### PR TITLE
Feature/landslide

### PR DESCRIPTION
**Pull request by Evelyn-M**
Hello both.

Here's my new pull request.

It's still about the landslide feature, no surprises :D. I removed the util function again, turned out after thorough re-consideration that there was no need for it.

So please review my module `hazard/landslide.py`, with the 3 functions it provides, as well as the corresponding tests and the tutorial.

A lot of the file changes are from deleting large test files, removing hard-coded file paths (both in `config` and `__init__()`)from previously uploaded files (which are now up to the user to download themselves).

Thanks for reviewing ;) !!

### CODE REVIEW

File: climada/hazard/landslide.py
Comment (ThomasRoosli) [inline climada/hazard/landslide.py:218]: The mathematics of fraction and/or frequency calculation should be checked. I think it might be `binom.rvs(n=n_years,p=self.fraction[i, j])/n_years`

File: climada/util/coordinates.py
Comment (ThomasRoosli) [inline climada/util/coordinates.py:37]: delete this line

**Review by ThomasRoosli — COMMENTED** [review]
Great work to integrate two different landslide datasets into CLIMADA that are very different from the event based mental model of a hazard, that CLIMADA generally uses.
As the two resulting hazards have a very different structure I would either split the current code into two classes: e.g. LandslidePoint, that reads in the historic dataset and LandslideGrid, that reads in the probabilistic dataset.
on the other hand: it is maybe worthwhile to invest a little more in a restructuring of the class to combine the strength of the existing code and produce a class where the historic and probabilistic Landslide hazard have a similar structure:
1. store the historic events on the same grid as the probabilistic events (res ca 500-1000 Meters), that would also solve the `assign_centroids(...,threshold=0.5)` workaround from the tutorial.
2. integrate `sample_events_from_probs` into the `.set_ls_prob` function and expand it to really create `n_years` events (each representing a year).

Please reach out if you have further questions.
I have already added a few other comments, but I will review again when the changes are implemented

**Comment by Evelyn-M** [comment]
Thanks for the feedback so far!
So I made quite some changes to the logic, and hence have not adjusted the unit tests yet.
As discussed I incorporated the following:
- `set_ls_hist` not comes in the same output format as the probabilistic one. It converts the point hazards into raster hazard sets with a user-specifiable resolution. A whole grid cell counts as affected if the point falls within.
- `set_ls_prob` by default produces sampled event-sets now. It is hence of the same logic as the historic one (only that it is properly probabilistic, and not a collection of single events)
- I updated the tutorial accordingly.
- There are 3 helper functions now in the landslide module. 2 for looking up the position of a point on a grid cell matrix, 1 for the binomial sampling from a probability matrix. -  please advice whether they should be transferred to `util`

Thanks :D

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: As is in the current code, they are still integer values (number of "successes" or landslide occurrences per grid cell over the whole time period). As the whole grid cell is either unaffected (0) or one to several times, this seemed to be rasonable in the fraction.

In `self.frequency`, I then divide the fraction value by `n_years`. 

So this should be consistent for the impact calculation, right?

**Comment by chahank** [comment]
I went through the tutorial and it is very very good. Clear and well written. I have only a few minor suggestions below:

- Remove the
```
import sys
path.climada = ...
...
```
It is not part of other tutorials and should not be necessary in general.

- Set the path to the file according to the default configuration defined in `climada.util.config`. Also, define the paths once as constant (capital letters) instead of defining it several times in the script.

- The import of `entity` is different on the latest develop branch. Probably you should once merge the develop branch into the feature/landslide branch and run the test and tutorial.

- Do not set `reference_year` when setting the exposure from litpop (done twice). Using the default makes it so that the user does not need to download extra night-light data for the arbitrary years 2020 and 2019 used in the tutorial.

- For the note on the historic model and how frequency are set to 1/size_of_set. Maybe add that for `impact.at_event` no frequency correction is made.

- The imports of the landslide module are not needed, so you can remove the lines `import climada.hazard.landslide`

- Could you add the resolution of the probabilistic data to the explanation? 

- I am not sure that the option `check_plots` is good implementation. However, I have to have a look at the code first for that.

I hope these are useful, and stay tuned for the proper code review!

File: climada/hazard/landslide.py
Comment (chahank) [inline climada/hazard/landslide.py:210]: Maybe the naming of this method should reflect the fact that it is only compatible with the UNEP & NGI dataset.

File: climada/hazard/landslide.py
Comment (chahank) [inline climada/hazard/landslide.py:237]: Why? Could I not want to do an impact function which gives different damages depending on how many occurences there were? (This would require the fraction and intensity to be swapped)

File: climada/hazard/landslide.py
Comment (chahank) [inline climada/hazard/landslide.py:247]:
```suggestion
            file, in case it is not scaled to [0,1]. Default is 1'000'000
```

File: climada/hazard/landslide.py
Comment (chahank) [inline climada/hazard/landslide.py:251]: Same as above. Not sure if having plots here is a good idea. The Method already makes several things a once (read data, process data and set data), hence adding also graphs is sub-optimal.

File: climada/hazard/landslide.py
Comment (chahank) [inline climada/hazard/landslide.py:286]: Is this compatible with the latest version of the develop branch (aka march release?)

File: climada/hazard/landslide.py
Comment (chahank) [inline climada/hazard/landslide.py:295]:
```suggestion

```

File: climada/hazard/test/test_landslide.py
Comment (chahank) [inline climada/hazard/test/test_landslide.py:31]: Please add a test for a country that crosses the Anti-meridian (e.g. the Fidji or the US). I am not sure that the bounding boxes in the code can deal with this.

File: climada/hazard/test/test_landslide.py
Comment (chahank) [inline climada/hazard/test/test_landslide.py:38]: If the plotting remains, please add a test with the plots.

File: climada/hazard/test/test_landslide.py
Comment (chahank) [inline climada/hazard/test/test_landslide.py:32]: Tests for `mapping_point2grid` and `mapping_grid2flattened` are missing.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:85]: Also, would it not be useful to allow for different kinds of distributions besides the binomial?

**Review by chahank — CHANGES_REQUESTED** [review]
This is a very clean and clear pull request. Several small comments and a few design questions. Most prominently, should two functions not be made as util functions, should plotting be included in data processing methods, and can the bbox as implemented deal with the anti-meridian?

Cool stuff! Is a publication / applied work planed on this?

**Comment by chahank** [comment]
Before I forget, please make sure that the module is compatible with the latest changes to how `crs` are handled. The easiest is probably to merge the develop branch into this branch or rebase your branch on develop (be careful in doing so) and see whether tests and tutorial still run smoothly.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:39]: Nope, this was exactly my question (see above in my comment to this pull request).

> There are 3 helper functions now in the landslide module. 2 for looking up the position of a point on a grid cell matrix, 1 for the binomial sampling from a probability matrix. - please advice whether they should be transferred to util

Would you advise to put it in the coordinates module?

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:86]: Yes, I agree. So, would it I) be useful to create a sampling util module? And II), do you think I should already include a kwarg for the type of distribution, in case more options will be implemented under a `sample_events_from_probs` parent function?

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:39]: Ah yes, I remember, and this confused me. I think this would be a good idea indeed. As a matter of fact, just this morning I came across an application where I needed a function like `mapping_point2grid`.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:86]: 1) Either a sampling util module, or maybe it would also fit in an engine module. Maybe we should a) open an issue on this for an open discussion and b) form a working sub-group at the developer's meeting.
2) I think this could be useful for this particular module, no? Maybe a poisson distribution could be appropriate for sampling the landslide events? In this case, I would say yes, adding a kwarg for the distribution is useful. However, if it is only for a possible future compatibility, I think we can leave it as is because we do not yet really know what the future implementation will look like. Right?

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:178]: correct!

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:208]: so i can just delete the return statement, right?

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:141]: Actually, picking up a comment by you further down, I think it's best to not put any frequency, at all. If one really wanted to have it for some purpose, it's easy to just set this yourself by (no. events / timespan passed in set)

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:208]: Yes. Or return None. Maybe @emanuel-schmid can comment on what is best for CLIMADA consistency?

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:173]: Adding multiple new empty columns to a dataframe is not super straight-forward, see [here](https://stackoverflow.com/questions/39050539/how-to-add-multiple-columns-to-pandas-dataframe-in-one-assignment)
This was basically solution #7 , i now changed it to #6 with is a bit shorter and still works

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:169]: I see. This is a bit confusing then to have a variable 'bounding box' that is actually an event * locations . This can maybe be made a bit more explicit, but it is not super important.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:252]: fair. i removed it

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: moved it.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: Interesting, I was not aware of this difficulty. In this case two single column additions would also work clean :D. But your call, I just didn't like the reindex which does more than needed.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: as a general remark: I'll open an issue, and for now leave it in the landslide module with two distribution options (binom and poisson)

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:108]: espg 4326

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: I changed it now to return a different object, generated initially from a copy of the prob_matrix, on which the trafos are applied

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:174]: mmh no, this should be included anyways, because impact = intensity*MDD*fraction

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:108]: I dont see how this is related. 1° at the equator is always ca 111km, irrespective of the projection, as long as a full circle is 360°.
And of course, one can manually reproject the gdf if one wanted to, for some reason.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: yes. changed it now.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: i changed this syntax. self.centroids.meta['crs'] now evaluates to CRS.from_epsg(4326). is this ok? it is read into this format when using the util.coordinates function read_raster

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:113]: Ok, then I probably misunderstood something. My concern is the following: with this method, your are dividing the surface in square cells of side lenght defined in ° (the raster). Then, you assign the landslide events to each raster cell. In this case, depending on the projection used, the cells and associated landslide events will vastly differ no?

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:180]: Yes, but only as direct addition. This means 2 events always imply twice the damage, which is not necessarily meaningful / desired.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: hmmm... in `util.coordinates` I find the line `CRS.from_user_input(DEF_CRS)`. Wouldn't this be more suitable (also thinking of future applications with flexible crs choices)?

File: climada/hazard/test/test_landslide.py
Comment (Evelyn-M) [reply climada/hazard/test/test_landslide.py:32]: I added them now

File: climada/hazard/test/test_landslide.py
Comment (Evelyn-M) [reply climada/hazard/test/test_landslide.py:31]: True that. it does not.
But in my opinion, this is nothing that has to do with the landslide module, but with the read_raster function, which does the reading of raster files incl bounding boxes of shapely (box) geometries and where this originates from.

I will open an issue on that. We can write sth where, if it a bbox crosses the anti-meridian, it actually reads in two files (left & right of the anti-meridian) and combines them

Comment (Evelyn-M) [reply climada/hazard/test/test_landslide.py:31]: alternatively, i could also just write a wrapper in my landslide module that does that, but i dont find this very clean, since it applies to _any_ raster files that are read in from a bbox

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:112]: correct. i believe this is the principle of a raster file.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: I re-changed my mind. climada forces you to put a non-nan frequency. Hence i still think putting a warning and a crappy estimate is better than putting 1 (which implies annual). I'll change it back.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: this just explicitly sets the crs in case it is `None` in the file that is read in. Aka it just "blindly" sets the crs to EPSG:4326

**Comment by Evelyn-M** [comment]
Alright.
Thanks a lot for the detailed review.
Tests are up and running, so is the tutorial (incl. the convention changes on crs format), and I incorporated the suggestions and / or added my comments on the discussion point. Basically the only things i did _not_ incorporate were including the data source names in the set_ls functions and changing back the order of fraction vs. intensity, since this was the whole point of re-defining intensity and fraction to match the physical background of how the landslides are represented in the data sets.

As such, I'd be ready to merge it. Please let me know if that's ok with you @chahank  @ThomasRoosli .

File: climada/hazard/test/test_landslide.py
Comment (chahank) [reply climada/hazard/test/test_landslide.py:34]: I agree with your point. It does not make sense to have a special solutions just in this module, but rather the `read_raster` method should be adapted. Opening an issue on this is certainly a good idea.

Comment (chahank) [reply climada/hazard/test/test_landslide.py:34]: However, don't you also have some potential `Landslide` specific issues in the part here 

```
        ls_gdf_bbox = ls_gdf_bbox.assign(row=np.nan, col=np.nan)
        ls_gdf_bbox[['col', 'row']] = ls_gdf_bbox.apply(
            lambda row: mapping_point2grid(row.geometry, bbox[-1], bbox[0], 
                                           res), axis = 1).tolist()
        ls_gdf_bbox['flat_ix'] = ls_gdf_bbox.apply(
            lambda row: mapping_grid2flattened(row.col, row.row, 
                                               self.centroids.shape), axis = 1)        
        self.intensity = sparse.csr_matrix(
            (np.ones(n_ev), (np.arange(n_ev), ls_gdf_bbox.flat_ix)), 
            shape=(n_ev, self.centroids.size))
        self.fraction = self.intensity.copy()
        self.frequency = self.intensity.copy()
```
?

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: OK, fair enough.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: Ok. If possible in the future, please do not close a comment if changes have been made that should also be reviewed (like in this case).

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:114]: I don't think so. The idea of a raster file is precisely the opposite. The information included in the raster allows to work with a regular raster on a mathematical level, but there is an implicit projection. Hence, the raster cell size is not independent on the projection. If instead of using degrees, one uses kilometers, the raster file still has the same mathematical regular grid representation, but the square cell side length is completely different.

**Comment by chahank** [comment]
There are 2-3 still open points, and I will now have another look at the test files.

File: climada/hazard/test/test_landslide.py
Comment (chahank) [reply climada/hazard/test/test_landslide.py:48]: If I am not mistaken, the test with `corr_fact = NotDefaultValue` is missing.

File: climada/util/coordinates.py
Comment (chahank) [inline climada/util/coordinates.py:830]: A bit more general, would allow not only tuples but also arrays of np.arrays
```suggestion
    if len(res) == 2:
        res_x, res_y = (abs(res) for res in res)
    else:
```

File: climada/util/coordinates.py
Comment (chahank) [inline climada/util/coordinates.py:837]: I still have some doubts about this one. 

Is the goal of this function to assign a geographical point (thus, given in a specific coordinate system) to a given grid cell of a raster? Then your function will only work if and only if the raster and the coordinate system are by chance defined in the same coordinate system, which is generally not case right?

File: climada/util/coordinates.py
Comment (chahank) [inline climada/util/coordinates.py:842]:
```suggestion
    - assumes concatenation along the row-axis (x-direction)
```

**Review by chahank — COMMENTED** [review]
Thanks for putting the general function to the util! I think they will be useful to many users/developers. 

I still am not 100% sure whether the `mapping_point2grid` is doing what it should be doing.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: Very nice for the reviewer is also if the changes are made in one commit that is referenced in the comments (for larger changes).

File: climada/util/coordinates.py
Comment (Evelyn-M) [reply climada/util/coordinates.py]: a float object has no length, hence len(res) would not work. that's why i do the type checking. i can add tuple _or_ list in the check

Comment (chahank) [reply climada/util/coordinates.py]: Aaa yes. Hmmm then the 'pythonic' way would probably be a try...except block ... (better ask for forgiveness than for permission)

File: climada/util/coordinates.py
Comment (Evelyn-M) [reply climada/util/coordinates.py:837]: Sure.
I can put this as a remark in the docstring. 
We have functions in climada to check beforehand whether projections are the same, in case of users needing to check for that.

File: climada/hazard/test/test_landslide.py
Comment (Evelyn-M) [reply climada/hazard/test/test_landslide.py:48]: added it now.

File: climada/hazard/test/test_landslide.py
Comment (Evelyn-M) [reply climada/hazard/test/test_landslide.py:34]: don't understand the question

Comment (Evelyn-M) [reply climada/hazard/test/test_landslide.py:34]: took this up in issue #172

File: climada/util/coordinates.py
Comment (Evelyn-M) [reply climada/util/coordinates.py]: commit 3ce3878 changes geometry to x and y

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: so, i now added in commit 1d0ca7a  the possibility to put in a gdf or a string of where the gdf is located, such that one may re-project the gdf into a more useful projection beforehand. also added some lengthy Note section in the docstring to explain how resolution affects the hazard extent

**Comment by Evelyn-M** [comment]
Alright. Incorporated the two things as discussed yesterday @chahank . Referred to the commits in the comments.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: I might have missed it, but I do not see the issue on this matter.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: Did you add tests for both use cases?

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:146]: I think I still do not understand why a bbox would contain events and not only locations.

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py]: But that is the point. The default use in CLIMADA is `DEF_CRS`. Why not us this default instead of implicitly defining a new one here?

**Comment by chahank** [comment]
Basically ready to merge, a few open minor comments that are not critical and one simple test is missing.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py:146]: it's not a bounding box. it's a geodataframe, cut to include only those entries that lie within the bounding box.I could re-name it to `gdf_cropped`or sth if that is more clar.

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: yep. that's the test i still wanted to add!

File: climada/hazard/landslide.py
Comment (chahank) [reply climada/hazard/landslide.py:146]: yes :D

File: climada/hazard/landslide.py
Comment (Evelyn-M) [reply climada/hazard/landslide.py]: changed this now

**Comment by Evelyn-M** [comment]
I added the last test and also incorporated 2 of the minor changes.

**Comment by chahank** [comment]
This is ready for merging. I am still not 100% convinced by the choice of how frequency / intensity / fraction are handled. However, the module definitely is clean and working!

**Pull request merged**
